src/datasources/accounts/counterfactual-safes/counterfactual-safes.datasource.ts

```diff
@@ -0,0 +1,145 @@
+import { Account } from '@/domain/accounts/entities/account.entity';
+import { ICounterfactualSafesDatasource } from '@/domain/interfaces/counterfactual-safes.datasource.interface';
+import { ILoggingService, LoggingService } from '@/logging/logging.interface';
+import { Inject, Injectable, NotFoundException } from '@nestjs/common';
+import postgres from 'postgres';
+
+@Injectable()
+export class CounterfactualSafesDatasource
+  implements ICounterfactualSafesDatasource
+{
+  private readonly defaultExpirationTimeInSeconds: number;
+
+  constructor(
+    @Inject(CacheService) private readonly cacheService: ICacheService,
+    @Inject('DB_INSTANCE') private readonly sql: postgres.Sql,
+    @Inject(LoggingService) private readonly loggingService: ILoggingService,
+    @Inject(IConfigurationService)
+    private readonly configurationService: IConfigurationService,
+  ) {
+    this.defaultExpirationTimeInSeconds =
+      this.configurationService.getOrThrow<number>(
+        'expirationTimeInSeconds.default',
+      );
+  }
+
+  // TODO: the repository calling this function should:
+  // - check the AccountDataSettings to see if counterfactual-safes is enabled.
+  // - check the AccountDataType to see if it's active.
+  async createCounterfactualSafe(
+    account: Account,
+    createCounterfactualSafeDto: CreateCounterfactualSafeDto,
+  ): Promise<CounterfactualSafe> {
+    const [counterfactualSafe] = await this.sql<CounterfactualSafe[]>`
+      INSERT INTO counterfactual_safes 
+      ${this.sql([this.mapCreationDtoToRow(account, createCounterfactualSafeDto)])}
+      RETURNING *`;
+    const { key } = CacheRouter.getCounterfactualSafesCacheDir(account.address);
+    await this.cacheService.deleteByKey(key);
+    return counterfactualSafe;
+  }
+
+  async getCounterfactualSafe(id: string): Promise<CounterfactualSafe> {
+    const cacheDir = CacheRouter.getCounterfactualSafeCacheDir(id);
+    const [counterfactualSafe] = await getFromCacheOrExecuteAndCache<
+      CounterfactualSafe[]
+    >(
+      this.loggingService,
+      this.cacheService,
+      cacheDir,
+      this.sql<CounterfactualSafe[]>`
+        SELECT * FROM counterfactual_safes WHERE id = ${id}`,
+      this.defaultExpirationTimeInSeconds,
+    );
+
+    if (!counterfactualSafe) {
+      throw new NotFoundException('Error getting Counterfactual Safe.');
+    }
+
+    return counterfactualSafe;
+  }
+
+  getCounterfactualSafesForAccount(
+    account: Account,
+  ): Promise<CounterfactualSafe[]> {
+    const cacheDir = CacheRouter.getCounterfactualSafesCacheDir(
+      account.address,
+    );
+    return getFromCacheOrExecuteAndCache<CounterfactualSafe[]>(
+      this.loggingService,
+      this.cacheService,
+      cacheDir,
+      this.sql<CounterfactualSafe[]>`
+        SELECT * FROM counterfactual_safes WHERE account_id = ${account.id}`,
+      this.defaultExpirationTimeInSeconds,
+    );
+  }
+
+  async deleteCounterfactualSafe(account: Account, id: string): Promise<void> {
```

Comment:
As above regarding predicted address vs. ID.

Reply:
Answered in https://github.com/safe-global/safe-client-gateway/pull/1773#discussion_r1684141115